Cameron carson/penalty position dribbling

CameronLyon opened the pull request:
## Description
Members Cameron and Rocio modified the penalty-shooting.cpp and hpp files to implement a "dribbling" function that creates a small kick, a follow function, and then the scheduled shoot once the ball is close enough to the enemy goal (3.5 meters). This behavior has not been tested on the blue side.

## Associated / Resolved Issue
[ClickUp Post](https://app.clickup.com/t/86b2ca7n9)

## Steps to Test
1. Check to see if the penalty shot functions properly using SSL ref controller
2. Check to see if the penalty shot functions on both blue side and on the other side of the field
3. Check to see if the position of the goalie modifies the directed shot of the ball (if placed on left side, ball should be kicked to the right side of the goal)

**Expected result:** A fully autonomous penalty shot that follows the following steps
1. Lines up behind the ball at a very small distance
2. Pushes the ball at a given speed and continues to follow the ball until the robot is 3.5 meters away from the goal
3. Lines itself back up again, calculates the best shot possible, and shoots it from now a closer distance from where the ball originated.

## Key Files to Review
penalty_player.cpp
penalty_player.hpp

CameronLyon committed:
1/26/2025

CameronLyon committed:
2/2/2025, modified behaviour of penalty shooting so that the ball gets kicked once shortly, follows it, then kicks to goal. Only works when ball is at center field.

rociopv06 committed:
Modified penalty behavior to work from the correct position (y=3). Follows ball by kicking with speed 0 and then kicks when closer to the goal. WARNING: Has only been tested on yellow side

rociopv06 committed:
quick edit to the .hpp states

jacksherling commented on soccer/src/soccer/strategy/agent/position/penalty_player.cpp:
```diff
@@ -18,19 +18,30 @@ PenaltyPlayer::State PenaltyPlayer::update_state() {
         }
-        case SHOOTING_START: {
+        case SMALL_KICK: {
+            if (distance_from_enemy_goal() < 3.5) {
```
Magic number (3.5) - let's make this a constant, so we can change later if we want to! See how other files do this with variables that start with k.

jacksherling commented on soccer/src/soccer/strategy/agent/position/penalty_player.hpp:
```diff
@@ -40,7 +40,7 @@ class PenaltyPlayer : public Position {
     std::optional<RobotIntent> derived_get_task(RobotIntent intent) override;
 
-    enum State { LINE_UP, SHOOTING_START, SHOOTING };
+    enum State { LINE_UP, SHOOTING_START, SHOOTING, SMALL_KICK, LINE_UP_2 };
```
I think we can come up with a more descriptive name than LINE_UP_2. Feel free to change LINE_UP as well

jacksherling commented on soccer/src/soccer/strategy/agent/position/penalty_player.cpp:
```diff
@@ -43,14 +54,38 @@ PenaltyPlayer::State PenaltyPlayer::update_state() {
+            intent.kick_speed = 0.0; 
+            // the point of making a 0 kick speed is to fake dribble since we cannot get the vaccum behavior to work
+
+            return intent;
```
looks like we're returning and breaking. Let's only do one and keep it consistent with the other cases

jacksherling commented on soccer/src/soccer/strategy/agent/position/penalty_player.cpp:
```diff
@@ -43,14 +54,38 @@ PenaltyPlayer::State PenaltyPlayer::update_state() {
+        }
+        case SMALL_KICK: { //less of a kick, more of a "follow" ball closely
+            
+            rj_geometry::Point center_goal{0,9};
```
There should be a field dimension for this (along the lines of their_goal_center)

jacksherling commented on soccer/src/soccer/strategy/agent/position/penalty_player.cpp:
```diff
@@ -18,19 +18,30 @@ PenaltyPlayer::State PenaltyPlayer::update_state() {
+        case SHOOTING_START: {
+            if (check_is_done()) {
+                return SHOOTING;
+            }
```
missing a break here

CameronLyon committed:
2/4/2025, added improvements following Jack's comments

CameronLyon commented:
Everything you've commented on has been resolved and additional comments/improvements were made.

CameronLyon committed:
quick change to a single line

github-actions[bot] committed:
Fix Code Style On cameron-carson/penalty-position-dribbling (#2337)

automated style fixes

Co-authored-by: CameronLyon <CameronLyon@users.noreply.github.com>

jacksherling commented on soccer/src/soccer/strategy/agent/position/penalty_player.cpp:
```diff
@@ -43,14 +56,41 @@ PenaltyPlayer::State PenaltyPlayer::update_state() {
     switch (latest_state_) {
-        case LINE_UP: {
+        case START: {  // First, gets the robot to the ball to begin penalty dribbling-shooting
+            SPDLOG_INFO("START");
```
Looks like some logging was left here!

rishiso replied:
+1

jacksherling commented on soccer/src/soccer/strategy/agent/position/penalty_player.cpp:
```diff
@@ -43,14 +56,41 @@ PenaltyPlayer::State PenaltyPlayer::update_state() {
+        case START: {  // First, gets the robot to the ball to begin penalty dribbling-shooting
+            SPDLOG_INFO("START");
+            double y_pos = last_world_state_->ball.position.y();
+            y_pos -= kRobotRadius + 0.3;  // added the 0.01 as a buffer space
```
What is this doing--Why 0.3? Also, comment says 0.01.

jacksherling commented on soccer/src/soccer/strategy/agent/position/penalty_player.cpp:
```diff
@@ -43,14 +56,41 @@ PenaltyPlayer::State PenaltyPlayer::update_state() {
+            break;
+        }
+        case SMALL_KICK: {  // less of a kick, more of a "follow" ball closely
+            // SPDLOG_INFO("SMALL_KICK");
```
Let's clean these comments up too

jacksherling commented on soccer/src/soccer/strategy/agent/position/penalty_player.cpp:
```diff
@@ -43,14 +56,41 @@ PenaltyPlayer::State PenaltyPlayer::update_state() {
+        }
+        case LINE_UP: {  // gets the robot behind the ball with a certain distance (usually
+                         // immediatly skipped if the robot is already close)
+            // SPDLOG_INFO("LINE_UP");
```
Same as above on cleaning up comments

jacksherling commented on soccer/src/soccer/strategy/agent/position/penalty_player.cpp:
```diff
@@ -43,14 +56,41 @@ PenaltyPlayer::State PenaltyPlayer::update_state() {
-            // } else {
-            y_pos -= kRobotRadius + 0.3;
-            // }
+            y_pos -= kRobotRadius + 0.1;
```
What is the 0.1 doing? Can we relate this to the radius as a ratio or something?

jacksherling replied:
do you mean ```kOwnBallRadius``` possibly?

jacksherling commented on soccer/src/soccer/strategy/agent/position/penalty_player.cpp:
```diff
@@ -124,11 +168,16 @@ double PenaltyPlayer::distance_from_their_robots(rj_geometry::Point tail,
     rj_geometry::Point their_goal_pos = field_dimensions_.their_goal_loc();
-    double goal_width = field_dimensions_.goal_width();  // 1.0 meters
+    double goal_width =
+        field_dimensions_.goal_width();  // 1.0 meters // BALL SEEMS TO OCCASIONALLY MISS SHOT ON
```
Extra double slash here. Let's change this comment to TODO format.

rishiso commented on soccer/src/soccer/strategy/agent/position/penalty_player.cpp:
```diff
@@ -14,26 +14,39 @@ std::optional<RobotIntent> PenaltyPlayer::derived_get_task(RobotIntent intent) {
-        case SHOOTING_START: {
+        case SMALL_KICK: {
+            if (distance_from_enemy_goal() < kDistanceToGoalThreshold) {
+                return LINE_UP;
+            }
+            break;
+        }
+        case LINE_UP: {
             if (check_is_done()) {
-                return SHOOTING;
+                return SHOOTING_START;
             }
             if (distance_to_ball() < kOwnBallRadius) {
+                return SHOOTING_START;
+            }
```
Combine this into one if statement

rishiso commented on soccer/src/soccer/strategy/agent/position/penalty_player.cpp:
```diff
@@ -43,14 +56,41 @@ PenaltyPlayer::State PenaltyPlayer::update_state() {
+            planning::LinearMotionInstant goal{target_pt, target_vel};
+            intent.motion_command =
+                planning::MotionCommand{"path_target", goal, planning::FaceBall{}};
+            break;
```
Nitpicking but you use `break` here but a `return` statement on the other cases. Either method works but please use one or the other, not both

rishiso commented on soccer/src/soccer/strategy/agent/position/penalty_player.cpp:
```diff
@@ -62,9 +102,12 @@ std::optional<RobotIntent> PenaltyPlayer::state_to_task(RobotIntent intent) {
-        case SHOOTING_START: {
+        case SHOOTING_START: {  // Positions the robot behind the ball at a certain angle so that it
+                                // has a straight shot towards the goal
+            // SPDLOG_INFO("SHOOTING_START");
```
Remove this log comment

rishiso commented on soccer/src/soccer/strategy/agent/position/penalty_player.cpp:
```diff
@@ -77,7 +120,9 @@ std::optional<RobotIntent> PenaltyPlayer::state_to_task(RobotIntent intent) {
-        case SHOOTING: {
+        case SHOOTING: {  // Kicks the ball with a now much higher speed (basically SMALL_KICK but
+                          // power set to 4)
+            // SPDLOG_INFO("SHOOTING");
```
Remove this log comment

rishiso commented on soccer/src/soccer/strategy/agent/position/penalty_player.cpp:
```diff
@@ -124,11 +168,16 @@ double PenaltyPlayer::distance_from_their_robots(rj_geometry::Point tail,
     // Ball location
     rj_geometry::Point ball_position = this->last_world_state_->ball.position;
 
+    // Iterates across 19 possible shot locations along the goal width in 0.05-meter increments.
+    // For each location, it calculates the clearance distance from opponent robots and updates the
+    // best shot position if a better (less obstructed) option is found.
```
I think an extended description like this would be better as a docstring for the function

rishiso reviewed (changes requested):
Please fix comments

rociopv06 committed:
Edits based on pull request comments.
Mainly cleaned up comments and fixed inconsistencies in variable naming- improved code readability

github-actions[bot] committed:
Fix Code Style On cameron-carson/penalty-position-dribbling (#2345)

automated style fixes

Co-authored-by: rociopv06 <rociopv06@users.noreply.github.com>

rishiso reviewed (approved):
LGTM
